pkg/kubectl/cmd/util/factory.go

```diff
@@ -767,6 +770,27 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 				return false, fmt.Errorf("the object is not a pod or does not have a pod template")
 			}
 		},
+		// UpdateSelectorForObject update the selector for the provided object
+		UpdateSelectorForObject: func(obj runtime.Object, fn func(*map[string]string, *unversioned.LabelSelector) error) error {
```

Comment:
Seems like your usage actually just needs to get back the ref to the selector and then you can do whatever you want to it.